// react-obsidian/dist/transformers/babel-plugin-obsidian/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = plugin;
const unmagler_1 = __importDefault(require("./unmagler"));
const providerArgumentsTransformer = {
    visitor: {
        Program(path) {
            path.traverse(internalVisitor);
        },
    },
};
const internalVisitor = {
    ClassMethod: {
        enter({ node }) {
            unmagler_1.default.saveClassMethod('provides', node);
            unmagler_1.default.saveClassMethod('Provides', node);
        },
    },
    ClassProperty: {
        enter({ node }) {
            unmagler_1.default.saveClassProperty('inject', node);
            unmagler_1.default.saveClassProperty('lateInject', node);
            unmagler_1.default.saveClassProperty('Inject', node);
            unmagler_1.default.saveClassProperty('LateInject', node);
        },
    },
    Identifier: {
        enter({ node }) {
            unmagler_1.default.saveIdentifier('inject', node);
            unmagler_1.default.saveIdentifier('Inject', node);
        },
    },
    TSParameterProperty: {
        enter({ node }) {
            unmagler_1.default.saveTSParameterProperty('inject', node);
            unmagler_1.default.saveTSParameterProperty('Inject', node);
        },
    },
};
function plugin() {
    return providerArgumentsTransformer;
}
