// ts-morph-extensions/dist/project/projectRegistry.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProjectRegistry = void 0;
const ts_morph_1 = require("ts-morph");
const path = __importStar(require("path"));
const fs = __importStar(require("fs"));
const logger_1 = require("../logger");
const tsconfigParser_1 = require("../tsConfig/tsconfigParser");
const os = __importStar(require("os"));
const fileSystem_1 = require("../utils/fileSystem");
class ProjectRegistry {
    constructor(logger = new logger_1.NoOpLogger(), tsconfigParser = new tsconfigParser_1.TsConfigParser(), options) {
        this.logger = logger;
        this.tsconfigParser = tsconfigParser;
        this.options = options;
        this.projects = new Map();
        this.tempFiles = new Set();
    }
    get(filePathOrUri) {
        const filePath = this.normalizeFilePath(filePathOrUri);
        const tsConfigPath = this.resolveTsConfigPath(filePath);
        this.ensureProject(tsConfigPath);
        return this.projects.get(tsConfigPath);
    }
    getSourceFile(filePathOrUri) {
        const filePath = this.normalizeFilePath(filePathOrUri);
        const project = this.get(filePathOrUri);
        return project.getSourceFile(filePath);
    }
    getSourceFileOrThrow(filePathOrUri) {
        const filePath = this.normalizeFilePath(filePathOrUri);
        const project = this.get(filePathOrUri);
        return project.getSourceFileOrThrow(filePath);
    }
    normalizeFilePath(filePathOrUri) {
        return filePathOrUri.startsWith('file://') ? filePathOrUri.slice(7) : filePathOrUri;
    }
    ensureProject(tsConfigPath) {
        if (!this.projects.has(tsConfigPath)) {
            const tsConfig = this.tsconfigParser.parse(tsConfigPath);
            this.logger.debug(`Parsed tsconfig: ${tsConfigPath}`);
            const tempTsConfigPath = this.saveConfigToTempFile(tsConfig);
            const project = new ts_morph_1.Project({ tsConfigFilePath: tempTsConfigPath });
            this.projects.set(tsConfigPath, project);
        }
    }
    saveConfigToTempFile(tsConfig) {
        const tempDir = path.join(os.tmpdir(), 'obsidian-language-server');
        const tempTsConfigPath = path.join(tempDir, `tsconfig-${Date.now()}.json`);
        try {
            (0, fileSystem_1.ensureDir)(tempDir);
            (0, fileSystem_1.writeFile)(tempTsConfigPath, tsConfig);
            this.tempFiles.add(tempTsConfigPath);
            return tempTsConfigPath;
        }
        catch (error) {
            this.logger.error(`Error writing tsconfig to temp file: ${error}`);
            throw error;
        }
    }
    resolveTsConfigPath(filePath) {
        if (this.options?.overrideTsConfigPath)
            return this.options.overrideTsConfigPath;
        return this.findClosestTsConfig(filePath);
    }
    findClosestTsConfig(filePath) {
        let dir = path.dirname(filePath);
        while (dir !== path.parse(dir).root) {
            const maybeTsConfigPath = path.join(dir, 'tsconfig.json');
            if (fs.existsSync(maybeTsConfigPath)) {
                return path.resolve(dir, maybeTsConfigPath);
            }
            dir = path.dirname(dir);
        }
        throw new Error(`Could not find tsconfig.json for file: ${filePath}`);
    }
    dispose() {
        for (const tempFile of this.tempFiles) {
            this.clearTempTsConfigFiles(tempFile);
        }
        this.tempFiles.clear();
        this.projects.clear();
    }
    clearTempTsConfigFiles(tempFile) {
        try {
            if (fs.existsSync(tempFile)) {
                fs.unlinkSync(tempFile);
                this.logger.debug(`Cleaned up temporary tsconfig: ${tempFile}`);
            }
        }
        catch (error) {
            this.logger.error(`Error cleaning up temporary tsconfig ${tempFile}: ${error}`);
        }
    }
}
exports.ProjectRegistry = ProjectRegistry;
