// react-obsidian/dist/src/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockModel = exports.testKit = exports.useObservers = exports.useObserver = exports.injectHookWithArguments = exports.injectHook = exports.injectComponent = exports.mockGraphs = exports.Model = exports.MediatorObservable = exports.Observable = exports.Obsidian = exports.LifecycleBound = exports.LateInject = exports.Inject = exports.Injectable = exports.Provides = exports.Singleton = exports.Graph = exports.GraphMiddleware = exports.memoize = exports.lifecycleBound = exports.lateInject = exports.inject = exports.injectable = exports.provides = exports.ObjectGraph = exports.singleton = exports.graph = void 0;
const Graph_1 = require("./decorators/Graph");
const Inject_1 = require("./decorators/inject/Inject");
const Injectable_1 = require("./decorators/injectable/Injectable");
const LateInject_1 = require("./decorators/lateInject/LateInject");
const LifecycleBound_1 = require("./decorators/LifecycleBound");
const Provides_1 = require("./decorators/provides/Provides");
const Singleton_1 = require("./decorators/singleton/Singleton");
const Obsidian_1 = __importDefault(require("./Obsidian"));
const reactAvailability_1 = require("./utils/reactAvailability");
__exportStar(require("./types"), exports);
var Graph_2 = require("./decorators/Graph");
Object.defineProperty(exports, "graph", { enumerable: true, get: function () { return Graph_2.graph; } });
var Singleton_2 = require("./decorators/singleton/Singleton");
Object.defineProperty(exports, "singleton", { enumerable: true, get: function () { return Singleton_2.singleton; } });
var ObjectGraph_1 = require("./graph/ObjectGraph");
Object.defineProperty(exports, "ObjectGraph", { enumerable: true, get: function () { return ObjectGraph_1.ObjectGraph; } });
var Provides_2 = require("./decorators/provides/Provides");
Object.defineProperty(exports, "provides", { enumerable: true, get: function () { return Provides_2.provides; } });
var Injectable_2 = require("./decorators/injectable/Injectable");
Object.defineProperty(exports, "injectable", { enumerable: true, get: function () { return Injectable_2.injectable; } });
var Inject_2 = require("./decorators/inject/Inject");
Object.defineProperty(exports, "inject", { enumerable: true, get: function () { return Inject_2.inject; } });
var LateInject_2 = require("./decorators/lateInject/LateInject");
Object.defineProperty(exports, "lateInject", { enumerable: true, get: function () { return LateInject_2.lateInject; } });
var LifecycleBound_2 = require("./decorators/LifecycleBound");
Object.defineProperty(exports, "lifecycleBound", { enumerable: true, get: function () { return LifecycleBound_2.lifecycleBound; } });
var Memoize_1 = require("./decorators/memoize/Memoize");
Object.defineProperty(exports, "memoize", { enumerable: true, get: function () { return Memoize_1.memoize; } });
var GraphMiddleware_1 = require("./graph/registry/GraphMiddleware");
Object.defineProperty(exports, "GraphMiddleware", { enumerable: true, get: function () { return GraphMiddleware_1.GraphMiddleware; } });
/**
 * @deprecated Uppercase decorators are deprecated in favor of lowercase decorators to align with common naming conventions. Please use `@graph` instead.
 */
exports.Graph = Graph_1.graph;
/**
 * @deprecated Uppercase decorators are deprecated in favor of lowercase decorators to align with common naming convention. Please use `@singleton` instead.
 */
exports.Singleton = Singleton_1.singleton;
/**
 * @deprecated Uppercase decorators are deprecated in favor of lowercase decorators to align with common naming convention. Please use `@provides` instead.
 */
exports.Provides = Provides_1.provides;
/**
 * @deprecated Uppercase decorators are deprecated in favor of lowercase decorators to align with common naming convention. Please use `@injectable` instead.
 */
exports.Injectable = Injectable_1.injectable;
/**
 * @deprecated Uppercase decorators are deprecated in favor of lowercase decorators to align with common naming convention. Please use `@inject` instead.
 */
exports.Inject = Inject_1.inject;
/**
 * @deprecated Uppercase decorators are deprecated in favor of lowercase decorators to align with common naming convention. Please use `@lateInject` instead.
 */
exports.LateInject = LateInject_1.lateInject;
/**
 * @deprecated Uppercase decorators are deprecated in favor of lowercase decorators to align with common naming convention. Please use `@lifecycleBound` instead.
 */
exports.LifecycleBound = LifecycleBound_1.lifecycleBound;
exports.Obsidian = new Obsidian_1.default();
var Observable_1 = require("./observable/Observable");
Object.defineProperty(exports, "Observable", { enumerable: true, get: function () { return Observable_1.Observable; } });
var MediatorObservable_1 = require("./observable/mediator/MediatorObservable");
Object.defineProperty(exports, "MediatorObservable", { enumerable: true, get: function () { return MediatorObservable_1.MediatorObservable; } });
var Model_1 = require("./model/Model");
Object.defineProperty(exports, "Model", { enumerable: true, get: function () { return Model_1.Model; } });
var mockGraphs_1 = require("../testkit/mockGraphs");
Object.defineProperty(exports, "mockGraphs", { enumerable: true, get: function () { return mockGraphs_1.mockGraphs; } });
// ============================================================================
// React-dependent exports (conditionally loaded)
// ============================================================================
const reactAvailable = (0, reactAvailability_1.isReactAvailable)();
exports.injectComponent = reactAvailable
    ? require('./injectors/components/InjectComponent').injectComponent
    : (0, reactAvailability_1.createReactRequiredError)('injectComponent');
exports.injectHook = reactAvailable
    ? require('./injectors/hooks/InjectHook').injectHook
    : (0, reactAvailability_1.createReactRequiredError)('injectHook');
exports.injectHookWithArguments = reactAvailable
    ? require('./injectors/hooks/InjectHook').injectHookWithArguments
    : (0, reactAvailability_1.createReactRequiredError)('injectHookWithArguments');
exports.useObserver = reactAvailable
    ? require('./observable/useObserver').useObserver
    : (0, reactAvailability_1.createReactRequiredError)('useObserver');
exports.useObservers = reactAvailable
    ? require('./observable/useObservers').useObservers
    : (0, reactAvailability_1.createReactRequiredError)('useObservers');
exports.testKit = reactAvailable
    ? require('../testkit/index').testKit
    : (0, reactAvailability_1.createReactRequiredError)('testKit');
exports.mockModel = reactAvailable
    ? require('../testkit/index').mockModel
    : (0, reactAvailability_1.createReactRequiredError)('mockModel');
