// react-obsidian/dist/src/Obsidian.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const GraphRegistry_1 = __importDefault(require("./graph/registry/GraphRegistry"));
const LateInjector_1 = __importDefault(require("./injectors/class/LateInjector"));
const ServiceLocatorFactory_1 = __importDefault(require("./graph/ServiceLocatorFactory"));
class Obsidian {
    registerGraph(key, generator) {
        GraphRegistry_1.default.registerGraphGenerator(key, generator);
    }
    obtain(keyOrGraph, props) {
        return ServiceLocatorFactory_1.default.fromGraph(keyOrGraph, props);
    }
    inject(target, keyOrGraph) {
        return LateInjector_1.default.inject(target, keyOrGraph);
    }
    addGraphMiddleware(middleware) {
        GraphRegistry_1.default.addGraphMiddleware(middleware);
    }
    clearGraphMiddlewares() {
        GraphRegistry_1.default.clearGraphMiddlewares();
    }
    clearGraphs() {
        GraphRegistry_1.default.clearGraphMiddlewares();
        GraphRegistry_1.default.clearAll();
    }
}
exports.default = Obsidian;
