// react-obsidian/dist/src/graph/properties/PropertyRetriever.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const ProvidedPropertiesStore_1 = __importDefault(require("../../ProvidedPropertiesStore"));
const CircularDependenciesDetector_1 = require("../CircularDependenciesDetector");
const reflect_1 = require("../../utils/reflect");
const retrieveFromPublicSubgraphs_1 = require("./retrieveFromPublicSubgraphs");
const retrieveFromPublicAndPrivateSubgraphs_1 = require("./retrieveFromPublicAndPrivateSubgraphs");
class PropertyRetriever {
    constructor(graph) {
        this.graph = graph;
        this.retrieveFromPublicSubgraphs = new retrieveFromPublicSubgraphs_1.RetrieveFromPublicSubgraphs(this.graph);
        this.retrieveFromPublicAndPrivateSubgraphs = new retrieveFromPublicAndPrivateSubgraphs_1.RetrieveFromPublicAndPrivateSubgraphs(this.graph);
    }
    retrieve(property, receiver, maybeDetector) {
        return this.retrieveInternal(property, this.retrieveFromPublicSubgraphs, receiver, maybeDetector);
    }
    retrieveAll(property, receiver, maybeDetector) {
        return this.retrieveInternal(property, this.retrieveFromPublicAndPrivateSubgraphs, receiver, maybeDetector);
    }
    retrieveInternal(property, retrieveStrategy, receiver, maybeDetector) {
        const mangledPropertyKey = ProvidedPropertiesStore_1.default.getMangledProperty(this.graph, property);
        const circularDependenciesDetector = maybeDetector !== null && maybeDetector !== void 0 ? maybeDetector : new CircularDependenciesDetector_1.CircularDependenciesDetector(this.graph.name);
        if (mangledPropertyKey
            && mangledPropertyKey in this.graph
            && circularDependenciesDetector.visit(this.graph.name, property)) {
            const proxiedGraph = new Proxy(this.graph, {
                get(graph, dependencyName) {
                    return graph.retrieveAll(dependencyName, receiver, circularDependenciesDetector);
                },
            });
            const resolved = reflect_1.Reflect.get(this.graph, mangledPropertyKey, receiver)(proxiedGraph);
            circularDependenciesDetector.clear();
            return resolved;
        }
        if (circularDependenciesDetector.hasCircularDependencies()) {
            throw new Error(`Could not resolve ${circularDependenciesDetector.firstDependencyName}`
                + ` from ${circularDependenciesDetector.graphName} because of a circular dependency:`
                + ` ${circularDependenciesDetector.getDependencies().join(' -> ')}`);
        }
        return retrieveStrategy.retrieve(property, circularDependenciesDetector, receiver);
    }
}
exports.default = PropertyRetriever;
