// eslint-plugin-obsidian/dist/rules/unresolvedProviderDependencies/graphHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphHandler = void 0;
class GraphHandler {
    constructor(context, classAdapter) {
        this.context = context;
        this.classAdapter = classAdapter;
    }
    handle(clazz) {
        if (clazz.isGraph) {
            const unresolvedDependency = this.findUnresolvedDependency(clazz);
            this.reportError(this.context, clazz, unresolvedDependency);
        }
    }
    findUnresolvedDependency(clazz) {
        const graph = this.classAdapter.classToGraph(clazz.node, this.context.currentFilePath);
        const providers = graph.resolveProviders();
        for (const provider of graph.getProviders()) {
            const unresolvedDep = provider.dependencies.find(dep => dep.isNotProvided(providers));
            if (unresolvedDep)
                return { provider, unresolvedDep };
        }
    }
    reportError(context, clazz, providerToUnresolvedDependency) {
        if (providerToUnresolvedDependency) {
            const { provider, unresolvedDep } = providerToUnresolvedDependency;
            const parameter = clazz.requireMethodParameter(provider.name, unresolvedDep.name);
            context.reportError(parameter.node, 'unresolved-provider-dependencies', { dependencyName: unresolvedDep.name });
        }
    }
}
exports.GraphHandler = GraphHandler;
