// eslint-plugin-obsidian/dist/rules/unresolvedProviderDependencies/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unresolvedProviderDependenciesGenerator = void 0;
const utils_1 = require("@typescript-eslint/utils");
const createRule_1 = require("./createRule");
const context_1 = require("../../dto/context");
const projectRegistry_1 = require("../../ts/projectRegistry");
const classAdapter_1 = require("../../ts/adapters/classAdapter");
const createRule = utils_1.ESLintUtils.RuleCreator((name) => `https://wix-incubator.github.io/obsidian/docs/documentation/meta/eslint#${name}`);
const unresolvedProviderDependenciesGenerator = () => {
    return createRule({
        create: (context) => {
            return (0, createRule_1.create)(new context_1.Context(context), new classAdapter_1.ClassAdapter(projectRegistry_1.projectRegistry));
        },
        name: 'unresolved-provider-dependencies',
        meta: {
            docs: {
                description: 'Dependencies must be defined in the graph or its subgraphs.',
                recommended: 'strict',
            },
            messages: {
                'unresolved-provider-dependencies': 'Dependency {{ dependencyName }} is unresolved.',
            },
            schema: [],
            type: 'problem',
        },
        defaultOptions: [],
    });
};
exports.unresolvedProviderDependenciesGenerator = unresolvedProviderDependenciesGenerator;
