// ts-morph-extensions/dist/dto/graph.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Graph = void 0;
const index_1 = require("../index");
const ts_morph_1 = require("ts-morph");
const objects_1 = require("../utils/objects");
const dedupeSet_1 = require("../utils/dedupeSet");
class Graph {
    constructor(node) {
        this.node = node;
    }
    get name() {
        return this.node.getName();
    }
    resolveProvider(name) {
        return this.hasProvider(name) ?
            this.requireProvider(name) :
            this.resolveProviderFromSubgraphs(name);
    }
    hasProvider(name) {
        return this.findProvider(name) !== undefined;
    }
    requireProvider(name) {
        return this.findProvider(name);
    }
    resolveProviderFromSubgraphs(name) {
        for (const subgraph of this.getAllSubgraphs()) {
            const provider = subgraph.resolveProvider(name);
            if (provider)
                return provider;
        }
    }
    findProvider(name) {
        return this.getProviders().find(provider => provider.name === name);
    }
    getProviders() {
        const providers = (0, index_1.getDecoratedMethods)(this.node, ['Provides', 'provides']);
        return providers.map(method => new index_1.Provider(method));
    }
    getSubgraphs() {
        return this.getSubgraphsFromDecorator()
            .map(graph => this.getGraphFromSubgraph(graph))
            .filter(objects_1.isDefined);
    }
    getPrivateSubgraphs() {
        return this.getPrivateSubgraphsFromDecorator()
            .map(graph => this.getGraphFromSubgraph(graph))
            .filter(objects_1.isDefined);
    }
    getAllSubgraphs() {
        return [...this.getSubgraphs(), ...this.getPrivateSubgraphs()];
    }
    getGraphFromSubgraph(graph) {
        const declaration = (0, index_1.getDefinition)(graph, ts_morph_1.SyntaxKind.ClassDeclaration);
        return declaration && new Graph(declaration);
    }
    getSubgraphsFromDecorator() {
        const graphDecorator = (0, index_1.getDecorator)(this.node, ['Graph', 'graph']);
        const subgraphsArg = graphDecorator?.getArgument(0, 'subgraphs');
        return ts_morph_1.Node.isArrayLiteralExpression(subgraphsArg) ? subgraphsArg.getElements() : [];
    }
    getPrivateSubgraphsFromDecorator() {
        const graphDecorator = (0, index_1.getDecorator)(this.node, ['Graph', 'graph']);
        const privateSubgraphsArg = graphDecorator?.getArgument(0, 'privateSubgraphs');
        return ts_morph_1.Node.isArrayLiteralExpression(privateSubgraphsArg) ? privateSubgraphsArg.getElements() : [];
    }
    resolveProviders(dedupeSet = new dedupeSet_1.DedupeSet()) {
        const ownProviders = this.getProviders().filter(provider => dedupeSet.dedupe(provider.name));
        const allSubgraphProviders = this.getAllSubgraphs().flatMap(subgraph => subgraph.resolveProviders(dedupeSet));
        const baseGraphProviders = this.getBaseGraph()?.getProviders().filter(provider => dedupeSet.dedupe(provider.name)) || [];
        return [...ownProviders, ...baseGraphProviders, ...allSubgraphProviders];
    }
    getBaseGraph() {
        if (this.node.getBaseClass())
            return new Graph(this.node.getBaseClass());
    }
}
exports.Graph = Graph;
