// ts-morph-extensions/dist/dto/provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Provider = void 0;
const parameter_1 = require("./parameter");
class Provider {
    constructor(node) {
        this.node = node;
    }
    get name() {
        return this.node.getName().replace(/^_/, '');
    }
    get type() {
        const returnType = this.node.getReturnType();
        if (returnType.getCallSignatures().length > 0)
            return returnType.getText();
        const symbol = returnType.getSymbol();
        return symbol ? symbol.getName() : returnType.getText();
    }
    get kind() {
        if (this.node.getReturnType().isClassOrInterface())
            return "class";
        if (this.node.getReturnType().getCallSignatures().length > 0)
            return "function";
        return this.node.getReturnType().getText();
    }
    get uri() {
        return this.sourceFile.getFilePath();
    }
    get sourceFile() {
        return this.node.getSourceFile();
    }
    get definition() {
        const range = this.range;
        return {
            uri: this.uri,
            range: {
                start: {
                    line: range.start.line,
                    character: range.start.column - 1
                },
                end: {
                    line: range.end.line,
                    character: range.end.column
                }
            }
        };
    }
    get dependencies() {
        return this.node.getParameters().map(param => new parameter_1.Parameter(param));
    }
    get range() {
        return {
            start: this.sourceFile.getLineAndColumnAtPos(this.node.getStart()),
            end: this.sourceFile.getLineAndColumnAtPos(this.node.getEnd())
        };
    }
    hasParameter(name) {
        return this.node.getParameters().some(param => param.getName() === name);
    }
}
exports.Provider = Provider;
