// ts-morph-extensions/dist/tsConfig/tsconfigParser.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TsConfigParser = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const jsonc = __importStar(require("jsonc-parser"));
const relativeToAbsolutePathConverter_1 = require("./relativeToAbsolutePathConverter");
/**
 * Parses a tsconfig file and returns a normalized tsconfig object.
 *
 * This class is responsible for:
 * - Reading a tsconfig file
 * - Converting relative paths to absolute paths
 * - Merging composite tsconfig files (motivation: https://github.com/dsherret/ts-morph/issues/876)
 * - Returning a normalized tsconfig object
 */
class TsConfigParser {
    constructor(relativeToAbsoluteConverter = new relativeToAbsolutePathConverter_1.RelativeToAbsoluteConverter()) {
        this.relativeToAbsoluteConverter = relativeToAbsoluteConverter;
    }
    parse(configPath) {
        const rootConfig = this.readJsonFile(configPath);
        const baseDir = path.dirname(configPath);
        const configWithoutExtends = this.handleConfigExtends(rootConfig, baseDir);
        const tsConfig = this.isCompositeConfigFile(configWithoutExtends)
            ? this.parseCompositeConfigFile(configWithoutExtends, baseDir)
            : configWithoutExtends;
        this.relativeToAbsoluteConverter.convert(tsConfig, baseDir);
        return tsConfig;
    }
    handleConfigExtends(config, baseDir) {
        if (!config.extends) {
            return config;
        }
        const baseConfigPath = require.resolve(config.extends, { paths: [baseDir] });
        const baseConfigDir = path.dirname(baseConfigPath);
        const baseConfig = this.readJsonFile(baseConfigPath);
        // Recursively handle extends in the base config
        const resolvedBaseConfig = this.handleConfigExtends(baseConfig, baseConfigDir);
        // Remove the extends property to prevent infinite loops
        const { extends: _, ...configWithoutExtends } = config;
        return this.mergeConfigs(resolvedBaseConfig, configWithoutExtends);
    }
    parseCompositeConfigFile(rootConfig, baseDir) {
        let mergedConfig = this.createEmptyConfig();
        for (const reference of rootConfig.references) {
            const referencePath = path.resolve(baseDir, reference.path);
            const referenceConfig = this.readJsonFile(referencePath);
            this.mergeConfig(mergedConfig, referenceConfig);
        }
        return mergedConfig;
    }
    createEmptyConfig() {
        return {
            compilerOptions: {},
            include: [],
            exclude: [],
            files: [],
        };
    }
    isCompositeConfigFile(rootConfig) {
        return Boolean(rootConfig.references && Array.isArray(rootConfig.references));
    }
    readJsonFile(filePath) {
        const raw = fs.readFileSync(filePath, 'utf-8');
        return jsonc.parse(raw);
    }
    mergeConfig(target, source) {
        target.compilerOptions = {
            ...target.compilerOptions,
            ...source.compilerOptions,
        };
        this.mergeArrayProperties(target, source);
    }
    mergeArrayProperties(target, source) {
        for (const key of ['include', 'exclude', 'files']) {
            if (Array.isArray(source[key])) {
                target[key] = Array.isArray(target[key])
                    ? [...new Set([...target[key], ...source[key]])]
                    : [...source[key]];
            }
        }
    }
    mergeConfigs(baseConfig, extendingConfig) {
        // Create a new config object
        const mergedConfig = { ...this.createEmptyConfig() };
        // Merge compiler options
        mergedConfig.compilerOptions = {
            ...baseConfig.compilerOptions,
            ...extendingConfig.compilerOptions,
        };
        // Merge array properties
        for (const key of ['include', 'exclude', 'files']) {
            mergedConfig[key] = [];
            if (Array.isArray(baseConfig[key])) {
                mergedConfig[key].push(...baseConfig[key]);
            }
            if (Array.isArray(extendingConfig[key])) {
                mergedConfig[key].push(...extendingConfig[key]);
            }
            // Remove duplicates
            if (mergedConfig[key].length > 0) {
                mergedConfig[key] = [...new Set(mergedConfig[key])];
            }
        }
        // Merge references
        if (baseConfig.references || extendingConfig.references) {
            mergedConfig.references = [
                ...(baseConfig.references || []),
                ...(extendingConfig.references || []),
            ];
        }
        // Copy any other properties from the extending config
        for (const key in extendingConfig) {
            if (!['compilerOptions', 'include', 'exclude', 'files', 'references'].includes(key)) {
                mergedConfig[key] = extendingConfig[key];
            }
        }
        return mergedConfig;
    }
}
exports.TsConfigParser = TsConfigParser;
