// react-obsidian/dist/src/utils/reactAvailability.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isReactAvailable = isReactAvailable;
exports.createReactRequiredError = createReactRequiredError;
exports.createReactRequiredClass = createReactRequiredClass;
let reactAvailableCache;
function isReactAvailable() {
    if (reactAvailableCache !== undefined) {
        return reactAvailableCache;
    }
    try {
        // eslint-disable-next-line @typescript-eslint/no-require-imports, global-require
        require('react');
        reactAvailableCache = true;
    }
    catch (_a) {
        reactAvailableCache = false;
    }
    return reactAvailableCache;
}
/**
 * Creates a function that throws an error when called if React is not available.
 * Used to provide helpful error messages for React-dependent features.
 */
function createReactRequiredError(featureName) {
    return function reactRequiredStub(..._args) {
        throw new Error(`${featureName} requires React to be installed. `
            + `Please install React: npm install react`);
    };
}
/**
 * Creates a class constructor that throws an error when instantiated if React is not available.
 * Used for features like Model that need to be constructable.
 */
function createReactRequiredClass(className) {
    return class ReactRequiredStub {
        constructor() {
            throw new Error(`${className} requires React to be installed. `
                + `Please install React: npm install react`);
        }
    };
}
